src/type.c

```diff
@@ -5,6 +5,59 @@
 static SEXP syms_vec_is_vector_dispatch = NULL;
 static SEXP fns_vec_is_vector_dispatch = NULL;
 
+// Defined below
+static SEXP vec_type_rec(SEXP x, bool dispatch);
+
+
+static SEXP lgl_type(SEXP x) {
+  if (vec_is_unspecified(x)) {
+    return vctrs_shared_empty_uns;
+  } else {
+    return vec_slice(x, vctrs_shared_empty_int);
+  }
+}
+
+static SEXP df_type(SEXP x) {
+  R_len_t n = Rf_length(x);
+  SEXP out = PROTECT(Rf_allocVector(VECSXP, n));
+
+  for (R_len_t i = 0; i < n; ++i) {
+    SET_VECTOR_ELT(out, i, vec_type(VECTOR_ELT(x, i)));
+  }
+
+  // FIXME: Should that be restored?
+  SEXP nms = PROTECT(Rf_getAttrib(x, R_NamesSymbol));
+  Rf_setAttrib(out, R_NamesSymbol, nms);
+
+  out = df_restore(out, x, vctrs_shared_empty_int);
+
+  UNPROTECT(2);
+  return out;
+}
+
+static SEXP vec_type_dispatch(SEXP x) {
+  SEXP proxy = PROTECT(vec_proxy(x));
+  SEXP out = PROTECT(vec_type_rec(proxy, false));
+  out = vec_restore(out, x, vctrs_shared_empty_int);
+  UNPROTECT(2);
+  return out;
+}
+
+static SEXP vec_type_rec(SEXP x, bool dispatch) {
+  switch (vec_typeof_impl(x, dispatch)) {
+  case vctrs_type_scalar:    return x;
```

Comment:
This is currently required because partial frames are not vectors. Maybe they should become vectors though, since you're planning for a partial vector type.

Reply:
Ok, that makes sense